src/http/retry_strategy.cc

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#ifdef SEASTAR_MODULE
+module;
+#endif
+
+#include <chrono>
+#include <coroutine>
+#include <gnutls/gnutls.h>
+
+#ifdef SEASTAR_MODULE
+module seastar;
+#else
+#include <seastar/http/exception.hh>
+#include <seastar/http/retry_strategy.hh>
+#include <seastar/util/short_streams.hh>
+
+#endif
+
+using namespace std::chrono_literals;
+
+namespace seastar {
+extern logger http_log;
+
+namespace http::experimental {
+logger rs_logger("default_http_retry_strategy");
+
+extern bool is_retryable_exception(std::exception_ptr ex);
+
+default_retry_strategy::default_retry_strategy(unsigned max_retries, unsigned scale_factor)
+    : _max_retries(max_retries), _scale_factor(scale_factor) {
+}
+
+default_retry_strategy::default_retry_strategy()
+    : default_retry_strategy(2, 25) /* Default overall retry time to fail is 50ms (0ms)+((1<<1)*25)ms) */ {
+}
+
+future<bool> default_retry_strategy::should_retry(std::exception_ptr error, unsigned attempted_retries) const {
+    if (attempted_retries >= get_max_retries()) {
+        rs_logger.warn("Retries exhausted. Retry# {}", attempted_retries);
+        co_return false;
+    }
+
+    co_return is_retryable_exception(error);
+}
+
+future<input_stream<char>> default_retry_strategy::analyze_reply(std::optional<reply::status_type> expected, const reply& rep, input_stream<char>&& in) const {
+    co_return std::move(in);
```

Comment:
Looking at how it's implemented in Scylla (https://github.com/scylladb/scylladb/pull/24498) it looks like the implementation of the analyze_reply should resolve into exceptional future if it reads from `in`. Is that correct?

Reply:
right